File: frontend/src/engine/capabilities/consolidateAllSheets.ts
/**
 * consolidateAllSheets – Merge data from ALL worksheets into one combined sheet.
 *
 * Iterates every worksheet in the workbook, reads used-range data from each,
 * and writes the combined result into a single output sheet.
 *
 * Options:
 *   outputSheetName  → name of the destination sheet (default "Combined")
 *   hasHeaders       → if true, only include headers from the first sheet
 *   excludeSheets    → sheet names to skip
 */

import { CapabilityMeta, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";

const meta: CapabilityMeta = {
  action: "consolidateAllSheets",
  description: "Merge data from all worksheets into one combined sheet",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: any,
  options: ExecutionOptions
): Promise<StepResult> {
  const {
    outputSheetName = "Combined",
    hasHeaders = true,
    excludeSheets = [],
  } = params;

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would consolidate all sheets into "${outputSheetName}"`,
    };
  }

  options.onProgress?.("Loading worksheets...");

  // Load all worksheet names
  const sheets = context.workbook.worksheets;
  sheets.load("items/name");
  await context.sync();

  const excludeSet = new Set<string>([
    outputSheetName,
    ...(excludeSheets as string[]),
  ]);

  // Collect qualifying sheets and load their used ranges
  const qualifying: { name: string; range: Excel.Range }[] = [];
  for (const sheet of sheets.items) {
    if (excludeSet.has(sheet.name)) continue;
    const usedRange = sheet.getUsedRangeOrNullObject(true);
    usedRange.load(["values", "isNullObject"]);
    qualifying.push({ name: sheet.name, range: usedRange });
  }
  await context.sync();

  type Row = (string | number | boolean | null)[];
  const combined: Row[] = [];
  let sheetCount = 0;

  for (let i = 0; i < qualifying.length; i++) {
    const { name, range } = qualifying[i];
    if (range.isNullObject) continue;

    const values = (range.values ?? []) as Row[];
    if (values.length === 0) continue;

    options.onProgress?.(`Reading sheet "${name}" (${values.length} rows)...`);
    sheetCount++;

    if (combined.length === 0) {
      // First sheet: include everything (headers + data)
      combined.push(...values);
    } else {
      // Subsequent sheets: skip header row if hasHeaders
      const startRow = hasHeaders ? 1 : 0;
      for (let r = startRow; r < values.length; r++) {
        combined.push(values[r]);
      }
    }
  }

  if (combined.length === 0) {
    return {
      stepId: "",
      status: "success",
      message: "No data found in any qualifying sheets.",
    };
  }

  options.onProgress?.(`Writing ${combined.length} rows to "${outputSheetName}"...`);

  // Create or reuse the output sheet
  const existing = context.workbook.worksheets.getItemOrNullObject(outputSheetName);
  existing.load("isNullObject");
  await context.sync();

  const outSheet = existing.isNullObject
    ? context.workbook.worksheets.add(outputSheetName)
    : existing;

  // Write combined data
  const outRange = outSheet
    .getRange("A1")
    .getResizedRange(combined.length - 1, combined[0].length - 1);
  outRange.values = combined as any;
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Combined ${sheetCount} sheets (${combined.length} total rows) into "${outputSheetName}"`,
    outputs: { outputRange: `${outputSheetName}!A1` },
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/consolidate_all_sheets.py
"""consolidateAllSheets — merge data from all worksheets into one combined sheet.

Port of `frontend/src/engine/capabilities/consolidateAllSheets.ts`. Iterates
every worksheet in the workbook, reads used-range data, and writes a combined
matrix into a destination sheet (created if missing).
"""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry


def _as_2d(raw: Any, shape: tuple[int, int]) -> list[list[Any]]:
    if raw is None:
        return []
    rows_, cols_ = shape
    if not isinstance(raw, list):
        return [[raw]]
    if raw and not isinstance(raw[0], list):
        if rows_ == 1:
            return [list(raw)]
        return [[v] for v in raw]
    return [list(r) for r in raw]


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    output_sheet_name = params.get("outputSheetName") or "Combined"
    has_headers = bool(params.get("hasHeaders", True))
    exclude_sheets = params.get("excludeSheets") or []

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": f"Would consolidate all sheets into {output_sheet_name!r}.",
        }

    try:
        book = ctx.workbook_handle
        exclude_set = {output_sheet_name, *exclude_sheets}

        combined: list[list[Any]] = []
        sheet_count = 0

        for sheet in book.sheets:
            if sheet.name in exclude_set:
                continue

            used = sheet.used_range
            # An empty sheet's used_range can still report shape (1, 1) with a
            # None value. Skip when there's no meaningful data.
            try:
                values = _as_2d(used.value, used.shape)
            except Exception:  # noqa: BLE001
                continue

            if not values:
                continue
            # Skip sheets whose entire used range is empty/None.
            if all(all(v is None or v == "" for v in row) for row in values):
                continue

            sheet_count += 1
            if not combined:
                # First non-empty sheet: include everything (headers + data).
                combined.extend(values)
            else:
                start_row = 1 if has_headers else 0
                for r in range(start_row, len(values)):
                    combined.append(values[r])

        if not combined:
            return {
                "status": "success",
                "message": "No data found in any qualifying sheets.",
                "outputs": {},
            }

        # Create or reuse the output sheet.
        try:
            out_sheet = book.sheets[output_sheet_name]
        except Exception:  # noqa: BLE001
            out_sheet = book.sheets.add(output_sheet_name)

        width = max(len(r) for r in combined)
        for row in combined:
            while len(row) < width:
                row.append(None)

        out_sheet.range("A1").resize(len(combined), width).value = combined
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"consolidateAllSheets failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": (
            f"Combined {sheet_count} sheets ({len(combined)} total rows) "
            f"into {output_sheet_name!r}."
        ),
        "outputs": {"outputRange": f"{output_sheet_name}!A1"},
    }


registry.register("consolidateAllSheets", handler, mutates=True)
